File: src/main/java/at/pria/koza/harmonic/viewer/AbstractTreeNode.java
/**
 * AbstractTreeNode.java
 * 
 * Created on 11.08.2013
 */

package at.pria.koza.harmonic.viewer;


import static java.util.Collections.*;

import java.util.Enumeration;
import java.util.LinkedList;
import java.util.List;

import javax.swing.tree.TreeNode;


/**
 * <p>
 * {@code AbstractTreeNode}
 * </p>
 * 
 * @version V0.0 11.08.2013
 * @author SillyFreak
 */
public abstract class AbstractTreeNode<E extends TreeNode> implements TreeNode {
    
    protected abstract List<E> getChildren();
    
    @Override
    public abstract TreeNode getParent();
    
    protected Object[] getPath() {
        LinkedList<Object> path = new LinkedList<>();
        for(TreeNode node = this; node != null; node = node.getParent())
            path.addFirst(node);
        return path.toArray();
    }
    
    @Override
    public TreeNode getChildAt(int childIndex) {
        List<E> list = getChildren();
        if(list == null) throw new IndexOutOfBoundsException();
        return list.get(childIndex);
    }
    
    @Override
    public int getChildCount() {
        List<E> list = getChildren();
        return list == null? 0:list.size();
    }
    
    @Override
    public int getIndex(TreeNode node) {
        List<E> list = getChildren();
        if(list == null) return -1;
        return list.indexOf(node);
    }
    
    @Override
    public boolean getAllowsChildren() {
        return getChildren() != null;
    }
    
    @Override
    public boolean isLeaf() {
        List<E> list = getChildren();
        return list == null || list.isEmpty();
    }
    
    @Override
    public Enumeration<E> children() {
        List<E> list = getChildren();
        if(list == null) list = emptyList();
        return enumeration(list);
    }
}

File: src/main/scala/at/pria/koza/harmonic/viewer/AbstractTreeNode.scala
/**
 * AbstractTreeNode.scala
 *
 * Created on 11.08.2013
 */

package at.pria.koza.harmonic.viewer

import scala.collection.{ immutable, mutable }
import scala.collection.JavaConversions._

import java.util.Enumeration

import javax.swing.tree.TreeNode

/**
 * <p>
 * {@code AbstractTreeNode}
 * </p>
 *
 * @version V0.0 11.08.2013
 * @author SillyFreak
 */
abstract class AbstractTreeNode[E <: TreeNode](val parent: TreeNode) extends TreeNode with Seq[E] {
  //abstract members

  protected def childSeq: Seq[E]

  //Seq

  def iterator: Iterator[E] = childSeq.iterator

  def apply(idx: Int): E = childSeq(idx)
  def length: Int = childSeq.length

  //TreeNode

  override def getParent(): TreeNode = parent

  def allowsChildren: Boolean = childSeq != null
  override def getAllowsChildren(): Boolean = allowsChildren

  def childAt(index: Int): E =
    if (!allowsChildren) throw new IndexOutOfBoundsException()
    else childSeq(index)
  override def getChildAt(childIndex: Int): E = childAt(childIndex)

  def childCount: Int = if (!allowsChildren) 0 else childSeq.length
  override def getChildCount(): Int = childCount

  def indexOf(node: TreeNode): Int = if (!allowsChildren) -1 else childSeq.indexOf(node)
  override def getIndex(node: TreeNode): Int = indexOf(node)

  def leaf: Boolean = !allowsChildren || childCount == 0
  override def isLeaf(): Boolean = leaf

  override def children(): Enumeration[E] = (if (!allowsChildren) Seq.empty else childSeq).iterator

  //additional members

  protected[viewer] def path: immutable.Seq[Object] = {
    var path = List[Object]()
    var node: TreeNode = this
    while (node != null) {
      path = node :: path
      node = node.getParent()
    }
    path
  }
  protected[viewer] def pathArray: Array[Object] = {
    val p = path
    val result = new Array[Object](p.length)
    p.copyToArray(result)
    result
  }
}
